stat/stat.go

```diff
@@ -135,6 +138,32 @@ func (s *StatterWithTimeout) doLstat(path string, infoCh chan fs.FileInfo, errCh
 	}
 
 	info, err := s.lstat(path)
+	if err == nil {
+		stat, ok := info.Sys().(*syscall.Stat_t)
+		if ok {
+			s.correctFutureTimes(stat)
+			s.correctZeroAtimes(stat)
+		}
+	}
+
 	infoCh <- info
 	errCh <- err
 }
+
+func (s *StatterWithTimeout) correctFutureTimes(stat *syscall.Stat_t) {
+	if stat.Atim.Sec > s.defTime {
+		stat.Atim.Sec = s.defTime
+	}
+
+	if stat.Mtim.Sec > s.defTime {
+		stat.Mtim.Sec = s.defTime
+	}
+}
+
+func (s *StatterWithTimeout) correctZeroAtimes(stat *syscall.Stat_t) {
+	if stat.Atim.Sec == 0 && stat.Mtim.Sec != 0 {
```

Comment:
This seems a bit schizophrenic. There's no correctZeroMtimes method, but you have logic to check the Mtim isn't 0 here.

I'd suggest, even if you checked current production data and found no 0 mtimes, you add a correctZeroMtimes method than does something similar to this, but comparing to Ctim if that exists, otherwise (the corrected) Atim.

Reply:
We're not correcting zero mtimes, as I don't think we care about older mtimes, that check is to see if we the mtime is valid as an atime, and using that as a reasonable approximation to the correct atime.

If the mtime is also 0, we just use now, as I don't think we care about newer atimes.

I haven't seen an entry in out database that has zero for both atime and mtime, so setting the atime to now should rarely happen.

Reply:
I'm not sure about he context of this, but if this is being done at a high enough level, the mtime matters because we'll categorise the file it in to a certain mtime-based age bucket.

So even if there are not currently 0 mtimes, I think it makes sense to assume it's a possibility (just as a code currently does) and explicitly correct the mtimes.

Reply:
Don't we only propagate latest mtimes, though? If so, then the zero mtimes are essentially ignored.

It's the reverse for atimes, which is why 0 atimes are causing a problem, as they propagate up the tree.

Reply:
The new age buckets Rosie and I implemented last week don't care about latest or oldest. If this code is a high enough level, operating immediately on the only stat we do for a file before passing the stat info on to the various summarisation methods (and I guess it should be), then we want the age bucket summarisations to work with sane seeming m and a time values.